contrib/hyperscan/matching/input_matchers/source/matcher.cc

```diff
@@ -95,12 +59,88 @@ bool Matcher::match(absl::optional<absl::string_view> input) {
       },
       &matched);
   if (err != HS_SUCCESS && err != HS_SCAN_TERMINATED) {
-    ENVOY_LOG_MISC(error, "unable to scan, error code {}.", err);
+    ENVOY_LOG_MISC(error, "unable to scan, error code {}", err);
   }
 
   return matched;
 }
 
+std::string Matcher::replaceAll(absl::string_view value, absl::string_view substitution) const {
+  // Find matched pair.
+  using Found = std::pair<unsigned long long, unsigned long long>;
+  std::vector<Found> founds;
+  hs_scratch_t* scratch_ = tls_->get()->scratch_;
+  hs_error_t err = hs_scan(
+      som_database_, value.data(), value.size(), 0, scratch_,
+      [](unsigned int, unsigned long long from, unsigned long long to, unsigned int,
+         void* context) -> int {
+        std::vector<Found>* founds = static_cast<std::vector<Found>*>(context);
+        founds->push_back({from, to});
+
+        // Continue searching.
+        return 0;
+      },
+      &founds);
+  if (err != HS_SUCCESS && err != HS_SCAN_TERMINATED) {
+    ENVOY_LOG_MISC(error, "unable to scan, error code {}", err);
+    return std::string(value);
+  }
+
+  // Sort founds. Make sure the longest length found in the front will appear first.
+  std::sort(founds.begin(), founds.end(), [](Found a, Found b) {
+    if (a.first == b.first) {
+      return a.second > b.second;
+    }
+    return a.first < b.first;
+  });
+
+  // Concatenate string and replace matched pair with substitution.
+  std::string result;
+  unsigned long long pos = 0;
+  for (auto& found : founds) {
+    if (found.first < pos) {
+      continue;
+    }
+
+    result += std::string(value.substr(pos, found.first - pos));
+    result += std::string(substitution);
+    pos = found.second;
+  }
+  result += std::string(value.substr(pos));
+  return result;
+}
+
+bool Matcher::match(absl::optional<absl::string_view> input) {
+  if (!input) {
+    return false;
+  }
+
+  return static_cast<Envoy::Regex::CompiledMatcher*>(this)->match(*input);
+}
+
+void Matcher::compile(const std::vector<const char*>& expressions,
+                      const std::vector<unsigned int>& flags, const std::vector<unsigned int>& ids,
+                      hs_database_t** database) {
+  hs_compile_error_t* compile_err;
+  hs_error_t err =
+      hs_compile_multi(expressions.data(), flags.data(), ids.data(), expressions.size(),
+                       HS_MODE_BLOCK, nullptr, database, &compile_err);
+  if (err != HS_SUCCESS) {
+    std::string compile_err_message(compile_err->message);
```

Comment:
why is this copy needed?

Reply:
We must call `hs_free_compile_error` to free the `compile_err` before the calling of `throw`. So we have to store these messages in advance if we want to show in the throw statements.